src/ui/public/vis_maps/choropleth_layer.js

```diff
@@ -0,0 +1,142 @@
+import KibanaMapLayer from './kibana_map_layer';
+import $ from 'jquery';
+import L from 'leaflet';
+import _ from 'lodash';
+import colorramps from 'ui/vislib/components/color/colormaps';
+
+
+export default class ChoroplethLayer extends KibanaMapLayer {
+
+  constructor(geojsonUrl) {
+    super();
+
+    this._metrics = null;
+    this._joinField = null;
+    this._colorRamp = colorramps['Yellow to Red'];
+
+    this._geojsonUrl = geojsonUrl;
+    this._leafletLayer = L.geoJson(null, {
+      onEachFeature: (feature, layer) => {
+        layer.on('click', () => {
+          this.emit('select', feature.properties[this._joinField]);
+        });
+      },
+      style: emptyStyle
+    });
+
+    this._loaded = false;
+    $.ajax({
+      dataType: 'json',
+      url: geojsonUrl,
+      success: (data) => {
+        this._leafletLayer.addData(data);
+        this._loaded = true;
+        this._setChoroplethStyle();
+      }
+    }).error(function (e) {
+      // notifier.fatal(e);
+      // console.error(e);
+    });
+  }
+
+  _setChoroplethStyle() {
+    if (!this._loaded || !this._metrics || !this._joinField) {
+      return;
+    }
+    const styleFunction = makeChoroplethStyleFunction(this._metrics, this._colorRamp, this._joinField);
+    this._leafletLayer.setStyle(styleFunction);
+  }
+
+
+  setChoroplethJoinField(joinfield) {
+    if (joinfield === this._joinField) {
+      return;
+    }
+    this._joinField = joinfield;
+    this._setChoroplethStyle();
+  }
+
+
+  setChoroplethMetrics(metrics) {
+    this._metrics = metrics;
+    this._setChoroplethStyle();
+  }
+
+  setChoroplethColorRamp(colorRamp) {
+    if (_.isEqual(colorRamp, this._colorRamp)) {
+      return;
+    }
+    this._colorRamp = colorRamp;
+    this._setChoroplethStyle();
+  }
+
+  equalsGeoJsonUrl(geojsonUrl) {
+    return this._geojsonUrl === geojsonUrl;
+  }
+
+
+}
+
+
+function makeChoroplethStyleFunction(data, colorramp, joinField) {
+
+  if (data.length === 0) {
+    return function () {
+      return emptyStyle();
+    };
+  }
+
+  let min = data[0].value;
+  let max = data[0].value;
+  for (let i = 1; i < data.length; i += 1) {
```

Comment:
I'd rather not. imo, it's better practice to use simple for loops for more mathematical code. forEach loops create new closures, allocate objects on the heap iso. stack, ... Not that it would matter here in this particular case, but functional constructs wouldn't really contribute here, and readability of for-loop vs forEach is pretty much the same imo.